### ISSUE TITLE
bazel-6.2.1 fails to build with gcc-13

### ISSUE THREAD

**Issue by pva** [bazelbuild/bazel#18642]
### Description of the bug:

I've tried to build bazel with gcc-13 and it fails with the error:

```
INFO: From Action external/googleapis/google_bytestream_bytestream_java_grpc_srcs.jar:
google/bytestream/bytestream.proto:19:1: warning: Import google/api/annotations.proto is unused.
ERROR: /var/tmp/portage/portage/dev-util/bazel-6.2.1/temp/bazel_b52QQ2mO/out/external/com_google_absl/absl/strings/BUILD.bazel:1078:11: Compiling absl/strings/internal/str_format/float_conversion.cc [for tool] failed: (Exit 1): gcc failed: error executing command (from target @com_google_absl//absl/strings:str_format_internal) 
  (cd /var/tmp/portage/portage/dev-util/bazel-6.2.1/temp/bazel_b52QQ2mO/out/execroot/io_bazel && \
  exec env - \
    LD_LIBRARY_PATH=/opt/openjdk-bin-11.0.18_p10/lib/:/opt/openjdk-bin-11.0.18_p10/lib/server/ \
    PATH=/opt/openjdk-bin-11.0.18_p10/bin:/usr/lib/portage/python3.11/ebuild-helpers/xattr:/usr/lib/portage/python3.11/ebuild-helpers:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/bin:/usr/lib/llvm/16/bin:/usr/lib/llvm/15/bin:/etc/eselect/wine/bin:/usr/lib64/opencascade/bin \
    PWD=/proc/self/cwd \
  /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x' -MD -MF bazel-out/k8-opt-exec-EDC14992/bin/external/com_google_absl/absl/strings/_objs/str_format_internal/float_conversion.d '-frandom-seed=bazel-out/k8-opt-exec-EDC14992/bin/external/com_google_absl/absl/strings/_objs/str_format_internal/float_conversion.o' '-DBAZEL_CURRENT_REPOSITORY="com_google_absl"' -iquote external/com_google_absl -iquote bazel-out/k8-opt-exec-EDC14992/bin/external/com_google_absl -g0 -g0 -Wall -Wextra -Wcast-qual -Wconversion-null -Wformat-security -Wmissing-declarations -Woverlength-strings -Wpointer-arith -Wundef -Wunused-local-typedefs -Wunused-result -Wvarargs -Wvla -Wwrite-strings -DNOMINMAX -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c external/com_google_absl/absl/strings/internal/str_format/float_conversion.cc -o bazel-out/k8-opt-exec-EDC14992/bin/external/com_google_absl/absl/strings/_objs/str_format_internal/float_conversion.o)
# Configuration: d29cf4d6a39fe475ae79548870d8579c7aba1d9ab9074e95e8cf80f39b57db19
# Execution platform: //:default_host_platform
In file included from external/com_google_absl/absl/strings/internal/str_format/float_conversion.h:18,
                 from external/com_google_absl/absl/strings/internal/str_format/float_conversion.cc:15:
external/com_google_absl/absl/strings/internal/str_format/extension.h:34:6: warning: elaborated-type-specifier for a scoped enum must not use the 'class' keyword
   34 | enum class FormatConversionChar : uint8_t;
      | ~~~~ ^~~~~
      |      -----
external/com_google_absl/absl/strings/internal/str_format/extension.h:34:33: error: found ':' in nested-name-specifier, expected '::'
   34 | enum class FormatConversionChar : uint8_t;
      |                                 ^
      |                                 ::
external/com_google_absl/absl/strings/internal/str_format/extension.h:34:12: error: 'FormatConversionChar' has not been declared
   34 | enum class FormatConversionChar : uint8_t;
      |            ^~~~~~~~~~~~~~~~~~~~
external/com_google_absl/absl/strings/internal/str_format/extension.h:35:6: warning: elaborated-type-specifier for a scoped enum must not use the 'class' keyword
```

The full build log can be found here: https://bugs.gentoo.org/906914

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Try to build bazel on system with gcc-13

### Which operating system are you running Bazel on?

Gentoo Linux

**Comment by fmeum** [comment]
@meteorcloudy I think that's a red herring, looks like this is missing: https://chromium.googlesource.com/external/github.com/abseil/abseil-cpp/+/b957f0ccd00481cd4fd663d8320aa02ae0564f18%5E%21/absl/strings/internal/str_format/extension.h

**Comment by meteorcloudy** [comment]
Hmm, it looks like https://github.com/abseil/abseil-cpp/commit/b957f0ccd00481cd4fd663d8320aa02ae0564f18 is in the release of 20220623.1, which is used by Bazel https://github.com/bazelbuild/bazel/blob/2f7625c51c07c3405bf800521e3d2d1e89ce26a9/distdir_deps.bzl#L238

**Comment by meteorcloudy** [comment]
But in general, I would suggest to try to upgrade abseil-cpp and see if that fixes the issue.

**Comment by ncaq** [comment]
It's pretty convoluted and I'm not making progress in reading it, but for now, if you bump the abseil-cpp version as in the following commit, `bazel build //src:bazel-dev` won't go through.
If you do `BAZEL_CXXOPTS='-std=c++14' bazel build //src:bazel-dev`, it will pass.
I have no idea why c++0x is specified since the bazel source code looks like c++14 is the standard.

[build: bump abseil-cpp to 20230802.0 closed #18642 · ncaq/bazel@d2af2e1](https://github.com/ncaq/bazel/commit/d2af2e18ec1b2f955e4a1971bccfe4fe8543056e)

**Comment by ncaq** [comment]
Understood.
The commit that defaults to c++14 is included from v7, so until then it builds with `-std=c++0x`.
I was using v6, which is not pre-release, making the build impossible.
[Bump c++ standard to c++14 per default · bazelbuild/bazel@978cd23](https://github.com/bazelbuild/bazel/commit/978cd23e80f202a03e99a87f2089a9a0622b5d5d)

**Comment by ncaq** [comment]
@pva 

I was trying my best to fix it, but I think this is a problem that will be automatically fixed once bazel v7, which is already pre-released, is released, so I think the bazel side is at the level of making a decision here.

**Comment by adamjstewart** [comment]
This was fixed by #20785, you can backport that PR to support older versions.

**Comment by meteorcloudy** [comment]
Indeed. Upgrading from 6.2.1 to 6.5.0 should be easy, please try the latest 6.x release!
